src/randomization.ts

```diff
@@ -0,0 +1,350 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+export {
+    randomizeStimuli,
+    randomizeStimuliConstraints,
+    randomShuffle,
+    randomShuffleConstraints,
+    stimuliMeetConstraints
+};
+
+const NOT_SHUFFLED_ERROR_MSG =
+    "Unable to shuffle according to the constraints, " +
+    "perhaps it is an idea to loosen the constraints.";
+
+
+/**
+ * Swaps two items in an array
+ * @param stimuli The array with objects/stimuli
+ * @param i1 a positive integer smaller than stimuli.length
+ * @param i2 a positive integer smaller than stimuli.length
+ */
+function swapItems<T>(stimuli: Array<T>, i1: number, i2: number) {
+    let temp = stimuli[i1];
+    stimuli[i1] = stimuli[i2];
+    stimuli[i2] = temp;
+}
+
+// A constraint a key (string) that is allowed a number of times in a row
+type ConstraintMapping = { [key:string]: number};
+
+type StringToUnknownMapping = {[key:string] : unknown};
+
+/**
+ * Checks whether an item may be appended to the randomized items
+ *
+ * @param randomized {array.<object>} A pre-existing list of stimuli
+ * @param constraints {object} The constraints used to examine whether an
+ *                             item can be appended to randomized. The keys
+ *                             need to be a key in the item parameter and
+ *                             the value is the maximum number allowable
+ *                             subsequent item[key] with the same value.
+ * @param item {object} The object which is intended to be appended. The
+ *                      constraints determine whether it fits.
+ *
+ * @return true if the item may be appended to randomized without violating
+ *         the constraints false otherwise.
+ */
+function allowPushItem (
+    randomized: Array<StringToUnknownMapping>,
+    constraints: ConstraintMapping,
+    item: StringToUnknownMapping
+) {
+    for (const [key, max] of Object.entries(constraints)) {
+        if (max < 1) {
+            throw new RangeError(
+                "max is < 1; it's impossible to fit less than one repeating items"
+            );
+        }
+        let value: any = item[key];
+        if (value === undefined) {
+            throw ReferenceError(
+                `There is no key "${key}" in item "${item}"`
+            );
+        }
+
+        // Select the relevant items, if max tells only 3 items are
+        // allowed in a row, it only makes sense to look at those
+        let selection = randomized.slice(-max);
+        // Push the item that we want to append to the selection
+        selection.push(item);
+
+        let count = 0;
+        selection.forEach(stimulus => count += Number(stimulus[key] === value));
```

Comment:
Okay, looking at the original code this is 100% my fault. (So maybe this should be fixed as a separate issue). 

But this code is not really that readable, and should be a `reduce` anyway. 
Something like this would be better:
```ts
const count = selection.reduce((accumulator, stimulus) => {
    if (stimulus[key] === value)
        return accumulator + 1;
    return accumulator;
}, 0);
```